In [69]:
# 30 points
# Question 1 (30 points)

# what is the Probability Distribution for each F1 driver to win the Bahrain Grand Prix? 
# What is the Probability Distribution for each F1 driver to win both the Bahrain and Saudi Arabian Grand Prix? 
# What is the probability for Ferrari to win both races? 
# What is the probability for Ferrari to win at least one race? Note that Ferrari, and each other racing team, has two drivers per race.

# 30 points
#  If Ferrari wins the first race, what is the probability that Ferrari wins the next one? 
# If Ferrari wins at least one of these two races, what is the probability Ferrari wins both races? 
# How about Red Bull, Mercedes, and Alpine-Renault?

# 40 points
#Ferrari wins one of these two races on a rainy day. What is the probability Ferrari wins both races,
# assuming races can be held on either rainy, sunny, cloudy, snowy or foggy days? 
# Assume that rain, sun, clouds, snow, and fog are the only possible weather conditions on race tracks 
# and there is an equal probability for each one of these weather events.

# You need to provide proof for your answers by using code, in the form of a function. 
# I think it's one in a million because Ferrari sucks is not a good answer. Use the counting framework in this workbook.

#Hint: Use BGP to denote the Probability Distribution of F1 driver wins at the Bahrain GP. 
# Write driver initials as keys and driver wins as values in a dictionary that you pass to our function ProbDist..
# https://people.math.ethz.ch/~jteichma/probability.html

In [70]:
class ProbDist(dict):
    """A Probability Distribution; an {outcome: probability} mapping."""
    def __init__(self, mapping=(), **kwargs):
        self.update(mapping, **kwargs)
        # Make probabilities sum to 1.0; assert no negative probabilities
        total = sum(self.values())
        for outcome in self:
            self[outcome] = self[outcome] / total
            assert self[outcome] >= 0

In [71]:
from fractions import Fraction
def p(event, space): 
    """The probability of an event, given a sample space of equiprobable outcomes. 
    event: a collection of outcomes, or a predicate that is true of outcomes in the event. 
    space: a set of outcomes or a probability distribution of {outcome: frequency} pairs."""
    # branch on the type of the first argument
    if is_predicate(event):
        # transform the mapping (untangible) 'event' into the collection (tangible) 'event'
        event = such_that(event, space)
        
    if isinstance(space, ProbDist):
        # if space is a dictionary of distinct probabilities, where each item does not count as the same amount
        return sum(space[o] for o in space if o in event)
    else:
        # space is not a dictionary but a collection, let's fall back to our original division
        return Fraction(len(event & space), len(space))

is_predicate = callable

def such_that(predicate, space): 
    """The outcomes in the sample pace for which the predicate is true.
    If space is a set, return a subset {outcome,...} with outcomes where predicate(element) is true;
    if space is a ProbDist, return a ProbDist {outcome: frequency,...} with outcomes where predicate(element) is true."""
    if isinstance(space, ProbDist):
        return ProbDist({o:space[o] for o in space if predicate(o)})
    else:
        return {o for o in space if predicate(o)}

### What is the Probability Distribution for each F1 driver to win the Bahrain Grand Prix?

In [72]:
# BGP denotes the Probability Distribution of each F1 driver wins at the Bahrain GP.

BGP = ProbDist(
    MV = 454, CL = 308, SP = 305, GR = 275, CS = 246, LH = 240, LN = 122, EO = 92, FA = 81, VB = 49, DR = 37, SV = 37, KM = 25, PG = 23, LS = 18,
    MS = 12, YT = 12, ZG = 6, AA = 4, NL = 2, NV = 2, NH = 0)
BGP


{'MV': 0.19319148936170213,
 'CL': 0.13106382978723405,
 'SP': 0.12978723404255318,
 'GR': 0.11702127659574468,
 'CS': 0.1046808510638298,
 'LH': 0.10212765957446808,
 'LN': 0.05191489361702128,
 'EO': 0.03914893617021276,
 'FA': 0.03446808510638298,
 'VB': 0.020851063829787235,
 'DR': 0.01574468085106383,
 'SV': 0.01574468085106383,
 'KM': 0.010638297872340425,
 'PG': 0.00978723404255319,
 'LS': 0.007659574468085106,
 'MS': 0.005106382978723404,
 'YT': 0.005106382978723404,
 'ZG': 0.002553191489361702,
 'AA': 0.001702127659574468,
 'NL': 0.000851063829787234,
 'NV': 0.000851063829787234,
 'NH': 0.0}

### What is the Probability Distribution for each F1 driver to win Saudi Arabian Grand Prix? 

In [73]:

SAGP = ProbDist(
    MV = 454, CL = 308, SP = 305, GR = 275, CS = 246, LH = 240, LN = 122, EO = 92, FA = 81, VB = 49, DR = 37, SV = 37, KM = 25, PG = 23, LS = 18,
    MS = 12, YT = 12, ZG = 6, AA = 4, NL = 2, NV = 2, NH = 0)
SAGP

{'MV': 0.19319148936170213,
 'CL': 0.13106382978723405,
 'SP': 0.12978723404255318,
 'GR': 0.11702127659574468,
 'CS': 0.1046808510638298,
 'LH': 0.10212765957446808,
 'LN': 0.05191489361702128,
 'EO': 0.03914893617021276,
 'FA': 0.03446808510638298,
 'VB': 0.020851063829787235,
 'DR': 0.01574468085106383,
 'SV': 0.01574468085106383,
 'KM': 0.010638297872340425,
 'PG': 0.00978723404255319,
 'LS': 0.007659574468085106,
 'MS': 0.005106382978723404,
 'YT': 0.005106382978723404,
 'ZG': 0.002553191489361702,
 'AA': 0.001702127659574468,
 'NL': 0.000851063829787234,
 'NV': 0.000851063829787234,
 'NH': 0.0}

### What is the Probability Distribution for each F1 driver to win both the Bahrain and Saudi Arabian Grand Prix?

In [74]:
# Since these are the latest driver standings available on the website and considering the fact that Saudi Arabia GP is just one week after the
# Bahrain GP, We assume that the two races are independent events i.e. the outcome of Bahrain GP does not affect the outcome of Saudi Arabia GP. 
# And we also consider the above driver standings and their probability distribution as our initial data and We also assume the probability 
# distribution of each driver remains the same.

# To get the Probability Distribution for each F1 driver to win both the Bahrain and Saudi Arabian Grand Prix, we multiply the two probabilties 
# to arrive at the conclusion of BGP and SAGP.  
# We take these two dictionaries from the result obtained above and store the resulting probability distribution in a dictionary named both_wins 

both_wins = {}
for key in BGP:
    if key in SAGP:
        both_wins[key] = BGP[key] * SAGP[key]
print(both_wins)


{'MV': 0.03732295156179267, 'CL': 0.017177727478497058, 'SP': 0.016844726120416476, 'GR': 0.013693979176097782, 'CS': 0.010958080579447716, 'LH': 0.010430058850158441, 'LN': 0.0026951561792666365, 'EO': 0.0015326392032593932, 'FA': 0.0011880488909008604, 'VB': 0.0004347668628338615, 'DR': 0.0002478949751018561, 'SV': 0.0002478949751018561, 'KM': 0.00011317338162064282, 'PG': 9.578995020371207e-05, 'LS': 5.866908103214123e-05, 'MS': 2.6075147125396105e-05, 'YT': 2.6075147125396105e-05, 'ZG': 6.518786781349026e-06, 'AA': 2.8972385694884564e-06, 'NL': 7.243096423721141e-07, 'NV': 7.243096423721141e-07, 'NH': 0.0}


This means that 'MV' has a 3.7% chance of winning both races, 'CL' has 1.7% chance of winning both races and so on...

In [75]:
# We now define the universal sample space BOTH as a joint distribution of Probability Distribution for each F1 driver to win both 
# the Bahrain and Saudi Arabian Grand Prix. 
# Outcome 'MV MV' denotes the probability of 'MV' winning the first race and second race as well. 'MV CL' denotes the probability of 'MV' winning
# the first race and 'CL' winning the seccond race, and so on...Here, we will use set comprehension to get the universal sample space named as BOTH

def joint(A, B, sep=''):
    """The joint distribution of two independent probability distributions. 
    Result is all entries of the form {a+sep+b: P(a)*P(b)}"""
    return ProbDist({a + sep + b: A[a] * B[b] 
                        for a in A 
                        for b in B})

BOTH = joint(BGP, SAGP, ' ')
BOTH

{'MV MV': 0.03732295156179264,
 'MV CL': 0.02532041647804435,
 'MV SP': 0.025073789044816643,
 'MV GR': 0.0226075147125396,
 'MV CS': 0.020223449524671786,
 'MV LH': 0.019730194658216373,
 'MV LN': 0.010029515617926657,
 'MV EO': 0.00756324128564961,
 'MV FA': 0.006658940697148027,
 'MV VB': 0.0040282480760525105,
 'MV DR': 0.0030417383431416913,
 'MV SV': 0.0030417383431416913,
 'MV KM': 0.0020552286102308724,
 'MV PG': 0.0018908103214124025,
 'MV LS': 0.001479764599366228,
 'MV MS': 0.0009865097329108186,
 'MV YT': 0.0009865097329108186,
 'MV ZG': 0.0004932548664554093,
 'MV AA': 0.0003288365776369396,
 'MV NL': 0.0001644182888184698,
 'MV NV': 0.0001644182888184698,
 'MV NH': 0.0,
 'CL MV': 0.02532041647804435,
 'CL CL': 0.017177727478497048,
 'CL SP': 0.017010411951109087,
 'CL GR': 0.015337256677229506,
 'CL CS': 0.013719873245812577,
 'CL LH': 0.013385242191036659,
 'CL LN': 0.006804164780443636,
 'CL EO': 0.005131009506564053,
 'CL FA': 0.004517519239474873,
 'CL VB': 0.00273282

### What is the probability for Ferrari to win both races?

In [76]:
# Ferrari team has 2 drivers named 'CL' and 'CS'. We would isolate those results and compute the resulting probability for Ferrari to win both races.
def ferrari_win_both(outcome):
    return outcome.startswith('CS CL') or outcome.startswith('CL CS') or outcome.startswith('CS CS') or outcome.startswith('CL CL')
p(ferrari_win_both, BOTH)

0.05557555454956991

Hence, the probability of Ferrari to win both the races is 0.05557555454956991 or 5.55%

In [77]:
# We would also compute the probability of the two Ferrari drivers winning both the races individually!
def CL_CS_both(outcome):
    return outcome.startswith('CL CL') or outcome.startswith('CS CS')
p(CL_CS_both, BOTH)

0.02813580805794476

Hence, the probability of the two Ferrari drivers winning both the races individually is 0.02813580805794476 or 2.81%

### What is the probability for Ferrari to win at least one race?

In [78]:
# First, we define a predicate function which would give us the probability for Ferrari winning one race.
# We then add this result to probability for Ferrari winning both the races and subtract the probability of the two Ferrari drivers winning both the races individually.
# We do the subtraction once because in the resultant set there would be duplicate entries of 'CL CL' and 'CS CS'.
# The final result would give us the probabilitiy for Ferrari winning at least one race

def ferrari_win_one(outcome):
    return outcome.startswith('CS') or outcome.startswith('CL') or outcome.endswith('CS') or outcome.endswith('CL') 
p(ferrari_win_one, BOTH)

ferrari_win_atleast_one =  p(ferrari_win_one, BOTH) + p(ferrari_win_both, BOTH) - p(CL_CS_both, BOTH)
print(ferrari_win_atleast_one)

0.4433535536441825


Hence, the probability for Ferrari to win at least one race is 0.4433535536441825 or 44.33%

### If Ferrari wins the first race, what is the probability that Ferrari wins the next one?

In [79]:
# We now define a predicate function(a function that returns true for an outcome that is in the event) named ferrari_win_first which would 
# return 'CL ..' in outcome and 'CS ..' in the outcome such that(the subset of universal sample space for which the predicate(ferrari_win_first) = true) 
# The resultant set would give us all possible combination of probabilities for Ferrari winning the first race

def ferrari_win_first(outcome):
    return outcome.startswith('CS') or outcome.startswith('CL')
such_that(ferrari_win_first, BOTH)

{'CL MV': 0.10740609877870805,
 'CL CL': 0.07286581150626009,
 'CL SP': 0.0721560795760043,
 'CL GR': 0.0650587602734465,
 'CL CS': 0.05819801828097397,
 'CL LH': 0.056778554420462396,
 'CL LN': 0.028862431830401724,
 'CL EO': 0.02176511252784392,
 'CL FA': 0.019162762116906065,
 'CL VB': 0.011592288194177743,
 'CL DR': 0.008753360473154621,
 'CL SV': 0.008753360473154621,
 'CL KM': 0.005914432752131501,
 'CL PG': 0.00544127813196098,
 'CL LS': 0.00425839158153468,
 'CL MS': 0.0028389277210231203,
 'CL YT': 0.0028389277210231203,
 'CL ZG': 0.0014194638605115602,
 'CL AA': 0.0009463092403410401,
 'CL NL': 0.00047315462017052007,
 'CL NV': 0.00047315462017052007,
 'CL NH': 0.0,
 'CS MV': 0.0857853905829941,
 'CS CL': 0.05819801828097397,
 'CS SP': 0.057631154466548894,
 'CS GR': 0.05196251632229818,
 'CS CS': 0.04648283278285584,
 'CS LH': 0.045349105154005684,
 'CS LN': 0.023052461786619557,
 'CS EO': 0.017383823642368845,
 'CS FA': 0.01530532298947692,
 'CS VB': 0.009258775635609496,
 

In [80]:
# Now for ferrari to win the second race after winning the first, there are only four combinations possible 'CL CS', 'CS CL', 'CS CS', and 'CL CL'
# We now isolate them as below and get the required probability for given the condition: If Ferrari wins the first race, 
# what is the probability that Ferrari wins the next one? 

def ferrari_win_second_after_winning_first(outcome):
    return outcome.startswith('CS CS') or outcome.startswith('CL CL') or outcome.startswith('CL CS') or outcome.startswith('CS CL')
p(ferrari_win_second_after_winning_first, such_that(ferrari_win_first, BOTH))

0.2357446808510639

Hence, the probability that Ferrari wins the second race, given the condition that Ferrari wins the first race is 0.2357446808510639 or 23.57%


### If Redbull Racing wins the first race, what is the probability that Redbull Racing wins the next one as well?

In [81]:
# We now define a predicate function(a function that returns true for an outcome that is in the event) named redbull_win_first which would 
# return 'MV ..' in outcome and 'SP ..' in the outcome such that(the subset of universal sample space for which the predicate(redbull_win_first) = true) 
# The resultant set would give us all possible combination of probabilities for Red Bull Racing winning the first race

def redbull_win_first(outcome):
    return outcome.startswith('MV') or outcome.startswith('SP')
such_that(redbull_win_first, BOTH)

{'MV MV': 0.11555854567880466,
 'MV CL': 0.07839654640764722,
 'MV SP': 0.0776329436828974,
 'MV GR': 0.0699969164353993,
 'MV CS': 0.06261542342948447,
 'MV LH': 0.06108821797998484,
 'MV LN': 0.03105317747315896,
 'MV EO': 0.023417150225660856,
 'MV FA': 0.020617273568244887,
 'MV VB': 0.01247217783758024,
 'MV DR': 0.009417766938580997,
 'MV SV': 0.009417766938580997,
 'MV KM': 0.006363356039581755,
 'MV PG': 0.005854287556415214,
 'MV LS': 0.004581616348498863,
 'MV MS': 0.003054410898999242,
 'MV YT': 0.003054410898999242,
 'MV ZG': 0.001527205449499621,
 'MV AA': 0.0010181369663330808,
 'MV NL': 0.0005090684831665404,
 'MV NV': 0.0005090684831665404,
 'MV NH': 0.0,
 'SP MV': 0.0776329436828974,
 'SP CL': 0.05266728337958678,
 'SP SP': 0.05215429035965574,
 'SP GR': 0.04702436016034534,
 'SP CS': 0.042065427634345294,
 'SP LH': 0.04103944159448321,
 'SP LN': 0.0208617161438623,
 'SP EO': 0.015731785944551897,
 'SP FA': 0.013850811538138086,
 'SP VB': 0.00837888599220699,
 'SP DR':

In [82]:
# Now for Red Bull Racing to win the second race after winning the first, there are only four combinations possible 'MV MV', 'SP SP', 'MV SP', and 'SP MV'
# We now isolate them as below and get the required probability for given the condition: If Red Bull Racing wins the first race, 
# what is the probability that Red Bull Racing wins the next one? 

def redbull_win_second_after_winning_first(outcome):
    return outcome.startswith('MV MV') or outcome.startswith('SP SP') or outcome.startswith('MV SP') or outcome.startswith('SP MV')
p(redbull_win_second_after_winning_first, such_that(redbull_win_first, BOTH))

0.3229787234042552

Hence, the probability that Red Bull Racing wins the second race, given the condition that Red Bull Racing wins the first race is 
0.3229787234042552 or 32.29%

### If Mercedes wins the first race, what is the probability that Mercedes wins the next one as well?

In [83]:
# We now define a predicate function(a function that returns true for an outcome that is in the event) named mercedes_win_first which would 
# return 'GR ..' in outcome and 'LH ..' in the outcome such that(the subset of universal sample space for which the predicate(mercedes_win_first) = true) 
# The resultant set would give us all possible combination of probabilities for Mercedes winning the first race

def mercedes_win_first(outcome):
    return outcome.startswith('GR') or outcome.startswith('LH')
such_that(mercedes_win_first, BOTH)

{'GR MV': 0.1031605040280934,
 'GR CL': 0.06998554017764926,
 'GR SP': 0.06930386283825657,
 'GR GR': 0.0624870894443297,
 'GR CS': 0.05589754183020039,
 'GR LH': 0.054534187151415,
 'GR LN': 0.027721545135302627,
 'GR EO': 0.020904771741375748,
 'GR FA': 0.018405288163602564,
 'GR VB': 0.011134063210080564,
 'GR DR': 0.008407353852509814,
 'GR SV': 0.008407353852509814,
 'GR KM': 0.005680644494939063,
 'GR PG': 0.005226192935343937,
 'GR LS': 0.004090064036356125,
 'GR MS': 0.0027267093575707503,
 'GR YT': 0.0027267093575707503,
 'GR ZG': 0.0013633546787853751,
 'GR AA': 0.0009089031191902501,
 'GR NL': 0.00045445155959512504,
 'GR NV': 0.00045445155959512504,
 'GR NH': 0.0,
 'LH MV': 0.09003098533360877,
 'LH CL': 0.06107828960958481,
 'LH SP': 0.060483371204296636,
 'LH GR': 0.054534187151415,
 'LH CS': 0.048783309233629425,
 'LH LH': 0.04759347242305309,
 'LH LN': 0.024193348481718658,
 'LH EO': 0.018244164428837018,
 'LH FA': 0.01606279694278042,
 'LH VB': 0.009717000619706673,
 '

In [84]:
# Now for Mercedes to win the second race after winning the first, there are only four combinations possible 'GR GR', 'LH LH', 'GR LH', and 'LH GR'
# We now isolate them as below and get the required probability for given the condition: If Mercedes wins the first race, 
# what is the probability that Mercedes wins the next one? 

def mercedes_win_second_after_winning_first(outcome):
    return outcome.startswith('GR GR') or outcome.startswith('LH LH') or outcome.startswith('GR LH') or outcome.startswith('LH GR')
p(mercedes_win_second_after_winning_first, such_that(mercedes_win_first, BOTH))

0.2191489361702128

Hence, the probability that Mercedes wins the second race, given the condition that Mercedes wins the first race is 
0.2191489361702128 or 21.91%

### If Alpine Renault wins the first race, what is the probability that Alpine Renault wins the next one as well?

In [85]:
# We now define a predicate function(a function that returns true for an outcome that is in the event) named alpine_renault_win_first which would 
# return 'EO ..' in outcome and 'FA ..' in the outcome such that(the subset of universal sample space for which the predicate(alpine_renault_win_first) = true) 
# The resultant set would give us all possible combination of probabilities for Alpine Renault winning the first race

def alpine_renault_win_first(outcome):
    return outcome.startswith('EO') or outcome.startswith('FA')
such_that(alpine_renault_win_first, BOTH)

{'EO MV': 0.10273767064321733,
 'EO CL': 0.06969868404870251,
 'EO SP': 0.06901980076251384,
 'EO GR': 0.06223096790062723,
 'EO CS': 0.05566842946747018,
 'EO LH': 0.05431066289509286,
 'EO LN': 0.02760792030500554,
 'EO EO': 0.020819087443118927,
 'EO FA': 0.018329848727093843,
 'EO VB': 0.011088427007748126,
 'EO DR': 0.008372893862993483,
 'EO SV': 0.008372893862993483,
 'EO KM': 0.005657360718238839,
 'EO PG': 0.005204771860779732,
 'EO LS': 0.004073299717131964,
 'EO MS': 0.0027155331447546427,
 'EO YT': 0.0027155331447546427,
 'EO ZG': 0.0013577665723773214,
 'EO AA': 0.0009051777149182143,
 'EO NL': 0.00045258885745910716,
 'EO NV': 0.00045258885745910716,
 'EO NH': 0.0,
 'FA MV': 0.09045381871848483,
 'FA CL': 0.06136514573853157,
 'FA SP': 0.060767433280039376,
 'FA GR': 0.054790308695117466,
 'FA CS': 0.04901242159635963,
 'FA LH': 0.047816996679375244,
 'FA LN': 0.02430697331201575,
 'FA EO': 0.018329848727093843,
 'FA FA': 0.016138236379289145,
 'FA VB': 0.0097626368220391

In [86]:
# Now for Alpine Renault to win the second race after winning the first, there are only four combinations possible 'EO EO', 'FA FA', 'EO FA', and 'FA EO'
# We now isolate them as below and get the required probability for given the condition: If Alpine Renault wins the first race, 
# what is the probability that Alpine Renault wins the next one? 

def alpine_renault_win_second_after_winning_first(outcome):
    return outcome.startswith('EO EO') or outcome.startswith('FA FA') or outcome.startswith('EO FA') or outcome.startswith('FA EO')
p(alpine_renault_win_second_after_winning_first, such_that(alpine_renault_win_first, BOTH))

0.07361702127659575

Hence, the probability that Alpine Renault wins the second race, given the condition that Alpine Renault wins the first race is 
0.07361702127659575 or 7.36%

###  If Ferrari wins at least one of these two races, what is the probability Ferrari wins both races?

In [87]:
# If Ferrari wins the first race, what is the probability that Ferrari wins the next one?
# The first probability is p(ferrari_win_first, BOTH) calculates the probability that a Ferrari wins the first race given the sample space BOTH.
# The second probability is p(ferrari_win_second, BOTH) calculates the probability that a Ferrari wins the second race given the sample space BOTH.
# The last p function calculates the conditional probability for Ferrari to win the second race given the prior event that Ferrari wins the first race.

def ferrari_win_first(outcome):
    return outcome.startswith('CS') or outcome.startswith('CL')
p(ferrari_win_first, BOTH)
def ferrari_win_second(outcome):
    return outcome.endswith('CS') or outcome.endswith('CL')
p(ferrari_win_second, BOTH)

p(ferrari_win_second, such_that(ferrari_win_first, BOTH))


0.2357446808510639

Hence, the conditional probability for Ferrari to win the second race given the prior event that Ferrari wins the first race is 0.2357446808510639 or 23.57%


###  If Red Bull wins at least one of these two races, what is the probability Red Bull wins both races?

In [88]:
def redbull_win_first(outcome):
    return outcome.startswith('MV') or outcome.startswith('SP')
p(redbull_win_first, BOTH)
def redbull_win_second(outcome):
    return outcome.endswith('MV') or outcome.endswith('SP')
p(redbull_win_second, BOTH)

p(redbull_win_second, such_that(redbull_win_first, BOTH))

0.3229787234042552

Hence, the conditional probability for Red Bull Racing to win the second race given the prior event that Red Bull Racing wins the first race is 0.3229787234042552 or 32.29% 

###  If Mercedes wins at least one of these two races, what is the probability Mercedes wins both races?

In [89]:
def mercedes_win_first(outcome):
    return outcome.startswith('GR') or outcome.startswith('LH')
p(mercedes_win_first, BOTH)
def mercedes_win_second(outcome):
    return outcome.endswith('GR') or outcome.endswith('LH')
p(mercedes_win_second, BOTH)

p(mercedes_win_second, such_that(mercedes_win_first, BOTH))

0.2191489361702128

Hence, the conditional probability for Mercedes to win the second race given the prior event that Mercedes wins the first race is 0.2191489361702128 or 21.91%

###  If Alpine Renault wins at least one of these two races, what is the probability Alpine Renault wins both races?

In [90]:
def alpine_renault_win_first(outcome):
    return outcome.startswith('EO') or outcome.startswith('FA')
p(alpine_renault_win_first, BOTH)
def alpine_renault_win_second(outcome):
    return outcome.endswith('EO') or outcome.endswith('FA')
p(alpine_renault_win_second, BOTH)

p(alpine_renault_win_second, such_that(alpine_renault_win_first, BOTH))

0.07361702127659575

Hence, the conditional probability for Alpine Renault to win the second race given the prior event that Alpine Renault wins the first race is 0.07361702127659575 or 7.36%

### Ferrari wins one of these two races on a rainy day. What is the probability Ferrari wins both races, assuming races can be held on either rainy, sunny, cloudy, snowy or foggy days? Assume that rain, sun, clouds, snow, and fog are the only possible weather conditions on race tracks and there is an equal probability for each one of these weather events

In [91]:
# Assume that rain, sun, clouds, snow, and fog are the only possible weather conditions on race tracks and there is an equal probability for 
# each one of these weather events, which makes this an equi-probable event. Hence, we can use set comprehension.
# weather denotes the Probability Distribution of each weather condition.

weather = ProbDist(
    rain=0.20,
     sun=0.20,
  clouds=0.20,
    snow=0.20,
     fog=0.20)

weather

{'rain': 0.2, 'sun': 0.2, 'clouds': 0.2, 'snow': 0.2, 'fog': 0.2}

In [92]:
# We now define the a sample space JBGP as a joint distribution of BGP and weather. 
# Joining BGP and weather

def joint(A, B, sep=''):
    """The joint distribution of two independent probability distributions. 
    Result is all entries of the form {a+sep+b: P(a)*P(b)}"""
    return ProbDist({a + sep + b: A[a] * B[b]
                    for a in A
                    for b in B})

JBGP = joint(BGP, weather, ' ')
JBGP

{'MV rain': 0.03863829787234048,
 'MV sun': 0.03863829787234048,
 'MV clouds': 0.03863829787234048,
 'MV snow': 0.03863829787234048,
 'MV fog': 0.03863829787234048,
 'CL rain': 0.026212765957446843,
 'CL sun': 0.026212765957446843,
 'CL clouds': 0.026212765957446843,
 'CL snow': 0.026212765957446843,
 'CL fog': 0.026212765957446843,
 'SP rain': 0.02595744680851067,
 'SP sun': 0.02595744680851067,
 'SP clouds': 0.02595744680851067,
 'SP snow': 0.02595744680851067,
 'SP fog': 0.02595744680851067,
 'GR rain': 0.023404255319148966,
 'GR sun': 0.023404255319148966,
 'GR clouds': 0.023404255319148966,
 'GR snow': 0.023404255319148966,
 'GR fog': 0.023404255319148966,
 'CS rain': 0.020936170212765986,
 'CS sun': 0.020936170212765986,
 'CS clouds': 0.020936170212765986,
 'CS snow': 0.020936170212765986,
 'CS fog': 0.020936170212765986,
 'LH rain': 0.02042553191489364,
 'LH sun': 0.02042553191489364,
 'LH clouds': 0.02042553191489364,
 'LH snow': 0.02042553191489364,
 'LH fog': 0.02042553191489

In [93]:
# We now define the a sample space JSAGP as a joint distribution of JSAGP and weather. 
# Joining JSAGP and weather

def joint(A, B, sep=''):
    """The joint distribution of two independent probability distributions. 
    Result is all entries of the form {a+sep+b: P(a)*P(b)}"""
    return ProbDist({a + sep + b: A[a] * B[b]
                    for a in A
                    for b in B})

JSAGP = joint(SAGP, weather, ' ')
JSAGP

{'MV rain': 0.03863829787234048,
 'MV sun': 0.03863829787234048,
 'MV clouds': 0.03863829787234048,
 'MV snow': 0.03863829787234048,
 'MV fog': 0.03863829787234048,
 'CL rain': 0.026212765957446843,
 'CL sun': 0.026212765957446843,
 'CL clouds': 0.026212765957446843,
 'CL snow': 0.026212765957446843,
 'CL fog': 0.026212765957446843,
 'SP rain': 0.02595744680851067,
 'SP sun': 0.02595744680851067,
 'SP clouds': 0.02595744680851067,
 'SP snow': 0.02595744680851067,
 'SP fog': 0.02595744680851067,
 'GR rain': 0.023404255319148966,
 'GR sun': 0.023404255319148966,
 'GR clouds': 0.023404255319148966,
 'GR snow': 0.023404255319148966,
 'GR fog': 0.023404255319148966,
 'CS rain': 0.020936170212765986,
 'CS sun': 0.020936170212765986,
 'CS clouds': 0.020936170212765986,
 'CS snow': 0.020936170212765986,
 'CS fog': 0.020936170212765986,
 'LH rain': 0.02042553191489364,
 'LH sun': 0.02042553191489364,
 'LH clouds': 0.02042553191489364,
 'LH snow': 0.02042553191489364,
 'LH fog': 0.02042553191489

In [94]:
# We now define the a joint sample space which would be our universal sample space named J_total as a joint distribution of JBGP and JSAGP. 
# Joining JBGP and JSAGP

def joint(A, B, sep=''):
    """The joint distribution of two independent probability distributions. 
    Result is all entries of the form {a+sep+b: P(a)*P(b)}"""
    return ProbDist({a + sep + b: A[a] * B[b]
                    for a in A
                    for b in B})

J_total = joint(JBGP, JSAGP, ' ')
J_total

{'MV rain MV rain': 0.0014929180624716556,
 'MV rain MV sun': 0.0014929180624716556,
 'MV rain MV clouds': 0.0014929180624716556,
 'MV rain MV snow': 0.0014929180624716556,
 'MV rain MV fog': 0.0014929180624716556,
 'MV rain CL rain': 0.0010128166591217398,
 'MV rain CL sun': 0.0010128166591217398,
 'MV rain CL clouds': 0.0010128166591217398,
 'MV rain CL snow': 0.0010128166591217398,
 'MV rain CL fog': 0.0010128166591217398,
 'MV rain SP rain': 0.0010029515617926317,
 'MV rain SP sun': 0.0010029515617926317,
 'MV rain SP clouds': 0.0010029515617926317,
 'MV rain SP snow': 0.0010029515617926317,
 'MV rain SP fog': 0.0010029515617926317,
 'MV rain GR rain': 0.0009043005885015534,
 'MV rain GR sun': 0.0009043005885015534,
 'MV rain GR clouds': 0.0009043005885015534,
 'MV rain GR snow': 0.0009043005885015534,
 'MV rain GR fog': 0.0009043005885015534,
 'MV rain CS rain': 0.0008089379809868441,
 'MV rain CS sun': 0.0008089379809868441,
 'MV rain CS clouds': 0.0008089379809868441,
 'MV rain 

### What is the probability Ferrari wins both races, assuming races can be held on either rainy, sunny, cloudy, snowy or foggy days?

In [95]:
# The ferrari_wins_atleast_one_race_on_rainy_days function takes a single argument outcome, 
# which is a string. The function returns True if either the string 'CL rain' or 
# the string 'CS rain' is contained in outcome and False otherwise.
# The p function calculates the probability that Ferrari wins at least one race on rainy days, 
# given the joint distribution of the outcomes J_total.
# We now calculate the probability of Ferrari wins one of these two races on a rainy day.

def ferrari_wins_atleast_one_race_on_rainy_days(outcome):
    return 'CL rain' in outcome or 'CS rain' in outcome

p(ferrari_wins_atleast_one_race_on_rainy_days, J_total)

0.09207485015843989

In [96]:
# The function both_races_ferrari takes a string argument outcome and returns a boolean value indicating whether Ferrari won both races.
# We use .split(' ') method to split the outcome string into a list of substrings separated by spaces, 
# and then indexing is used to retrieve the first and third elements of the list and assign them to win_BGP_ferrari and win_SAGP_ferrari respectively.
# The function then returns True if either the string 'CL' is contained in win_BGP_ferrari and the string 'CS' is contained in win_SAGP_ferrari 
# or the string 'CS' is contained in win_BGP_ferrari and the string 'CL' is contained in win_SAGP_ferrari. 
# Thus the probability of Ferrari winning both races, given the prior event that Ferrari wins at least one race on a rainy day is computed.

def both_races_ferrari(outcome):
    win_BGP_ferrari = outcome.split(' ')[0]
    win_SAGP_ferrari = outcome.split(' ')[2]
    
    return (('CL' in win_BGP_ferrari and 'CS' in win_SAGP_ferrari) or
            ('CS' in win_BGP_ferrari and 'CL' in win_SAGP_ferrari) or
            ('CL' in win_BGP_ferrari and 'CL' in win_SAGP_ferrari) or
            ('CS' in win_BGP_ferrari and 'CS' in win_SAGP_ferrari)) 

p(both_races_ferrari, such_that(ferrari_wins_atleast_one_race_on_rainy_days, J_total))

0.21729277433975333

Thus the probability of Ferrari winning both races, given the prior event that Ferrari wins one of the two races on a rainy day 
is 0.21729277433975333 or 21.72%